src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -190,24 +236,56 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
     // and stick it in a map.
     val qs = new QuerySchema(z)
 
-    val whereClause = z.bodies(0).zipWithIndex flatMap {
-      case (Atom(relName, terms),body_index) =>
-        terms flatMap {
-          case Variable(varName, relName, index) =>
-            val canonical_body_index = qs.getBodyIndex(varName)
-
-            if (canonical_body_index != body_index) {
-              val real_attr_name1 = resolveName( Variable(varName, relName, index) )
-              val real_attr_name2 = resolveName( qs.getVar(varName))
-              Some(s"R${ body_index }.${ real_attr_name1 } = R${ canonical_body_index }.${ real_attr_name2 } ")
-            } else { None }
+    var whereClause = z.bodies(0).zipWithIndex flatMap {
+      case (Atom(relName, terms),bodyIndex) => {
+        terms flatMap { case ColumnExpr(expr, relName, index) =>
+          expr match {
+            // a simple variable
+            case VarExpr(varName) => {
+              val canonical_body_index = qs.getBodyIndex(varName)
+              if (canonical_body_index != bodyIndex) {
+                val real_attr_name1 = resolveName( Variable(varName, relName, index) )
+                val real_attr_name2 = resolveName( qs.getVar(varName))
+                Some(s"R${ bodyIndex }.${ real_attr_name1 } = R${ canonical_body_index }.${ real_attr_name2 } ")
+              } else { None }
+            }
+            // other expressions
+            case _ => {
+              val resolved = resolveExpr(expr, z, OriginalOnly, index, false)
+              val attr = schema(relName, index)
+              Some(s"R${bodyIndex}.${attr} = ${resolved}")
+            }
+          }
         }
+      }
+    }
+
+    // resolve conditions
+    val conditionStr = z.conditions(0) match {
+      case Some(c) => resolveCond(c, z)
+      case None    => ""
     }
-    val whereClauseStr = whereClause match {
-      case Nil => ""
-      case _ => s"""WHERE ${whereClause.mkString(" AND ")}"""
+
+    // handle group by
+    // map head terms, leaving out aggregation functions
+    val groupbyTerms = z.head.terms map { case ColumnExpr(expr, relName, index) =>
+      expr match {
+        case FuncExpr(f, args, agg) => if (agg) None else Some("")
+        case _ => Some(resolveExpr(expr, z, OriginalOnly, index, false))
+      }
+    }
+
+    val groupbyStr = if (groupbyTerms.size == z.head.terms.size) {
+      ""
+    } else {
+      s"\n        GROUP BY ${groupbyTerms.mkString(", ")}"
     }
 
+    var whereClauseStr = (whereClause match {
+      case Nil => if (conditionStr == "") "" else s"WHERE ${conditionStr}"
+      case _ => s"""WHERE ${whereClause.mkString(" AND ")} ${if (conditionStr == "") "" else s" AND (${conditionStr})"}"""
+    }) + groupbyStr
+
```

Comment:
Let's avoid creating brain teasers in code.  I'm sure this `(WHERE cond (AND cond)*)?` clause can be generated in a much simpler way: `whereClause` and compiled `conditions` can be concatenated and joined with `" AND "`, then prefixed with `"WHERE "` if non-empty.